File: GrooveOn.Models/RequestObjects/SongUpsertRequest.cs
using System;
using System.Collections.Generic;
using GrooveOn.Model.RequestObjects;

namespace GrooveOn.Model.Requests
{
    public class SongUpsertRequest
    {
        public string ExternalTrackId { get; set; } = string.Empty;
        public string? ExternalArtistId { get; set; }
        public string? ExternalAlbumId { get; set; }

        public string Source { get; set; } = "Deezer";

        public string Title { get; set; } = string.Empty;
        public string ArtistName { get; set; } = string.Empty;
        public string? ArtistPicture { get; set; }
        public string? AlbumTitle { get; set; }

        public int? DurationSeconds { get; set; }
        public string? PreviewUrl { get; set; }
        public string? CoverUrl { get; set; }
        public DateTime? ReleaseDate { get; set; }

        public List<GenreUpsertRequest> Genres { get; set; }
    }
}

File: UI/grooveon_desktop/lib/models/request/song_upsert_request.dart
import 'package:grooveon_desktop/models/request/genre_upsert_request.dart';
import 'package:json_annotation/json_annotation.dart';

part 'song_upsert_request.g.dart';

@JsonSerializable()
class SongUpsertRequest {
  final String externalTrackId;
  final String? externalArtistId;
  final String? externalAlbumId;
  final String source;
  final String title;
  final String artistName;
  final String? artistPicture;
  final String? albumTitle;
  final int? durationSeconds;
  final String? previewUrl;
  final String? coverUrl;
  final DateTime? releaseDate;
  final List<GenreUpsertRequest> genres;

  SongUpsertRequest({
    required this.externalTrackId,
    this.externalArtistId,
    this.externalAlbumId,
    this.source = "Deezer",
    required this.title,
    required this.artistName,
    this.artistPicture,
    this.albumTitle,
    this.durationSeconds,
    this.previewUrl,
    this.coverUrl,
    this.releaseDate,
    this.genres = const [],
  });

  factory SongUpsertRequest.fromJson(Map<String, dynamic> json) =>
      _$SongUpsertRequestFromJson(json);

  Map<String, dynamic> toJson() => _$SongUpsertRequestToJson(this);
}